### ISSUE TITLE
Unauthorized connection with same client_id makes the authorized client not receive messages anymore

### ISSUE THREAD

**Issue by llrosa** [chkr1011/MQTTnet#785]
### Describe the bug
If you use ConnectionValidationHandler in the MQTT Server to authorize and deny client connections, whenever you have an authorized client connected and a malicious client tries to connect using the same client_id of an authorized client and it gets denied by the authorization code, the original authenticated client stops receiving messages. The original client does not receive any events that it got disconnected or that there was a problem with its connection to the server.

The mechanism that denies connections for unauthorized clients works fine. The only issue there is that the unauthorized client gets a disconnected event, that is triggered by the server, even though that client never made to a successful connection. 

### Which project is your bug related to?
- MQTTnet.Server standalone

### To Reproduce
1. Using this version 3.0.5 of MQTTnet 
2. Run this test code I wrote at  Tests/MQTTnet.Core.Tests/Server_Tests.cs

```c#
private Dictionary<string, bool> _connected;
        private void ConnectionValidationHandler(MqttConnectionValidatorContext eventArgs)
        {
            if (_connected.ContainsKey(eventArgs.ClientId))
            {
                eventArgs.ReasonCode = MqttConnectReasonCode.BadUserNameOrPassword;
                return;
            }
            _connected[eventArgs.ClientId] = true;
            eventArgs.ReasonCode = MqttConnectReasonCode.Success;
            return;
        }

        [TestMethod]
        public async Task Same_Client_Id_Refuse_Connection()
        {
            using (var testEnvironment = new TestEnvironment())
            {
                testEnvironment.IgnoreClientLogErrors = true;

                _connected = new Dictionary<string, bool>();
                var options = new MqttServerOptionsBuilder();
                options.WithConnectionValidator(e => ConnectionValidationHandler(e));
                var server = await testEnvironment.StartServerAsync(options);

                var events = new List<string>();

                server.ClientConnectedHandler = new MqttServerClientConnectedHandlerDelegate(_ =>
                {
                    lock (events)
                    {
                        events.Add("c");
                    }
                });

                server.ClientDisconnectedHandler = new MqttServerClientDisconnectedHandlerDelegate(_ =>
                {
                    lock (events)
                    {
                        events.Add("d");
                    }
                });

                var clientOptions = new MqttClientOptionsBuilder()
                    .WithClientId("same_id");

                // c
                var c1 = await testEnvironment.ConnectClientAsync(clientOptions);

                c1.UseDisconnectedHandler(_ =>
                {
                    lock (events)
                    {
                        events.Add("x");
                    }
                });


                c1.UseApplicationMessageReceivedHandler(_ =>
                {
                    lock (events)
                    {
                        events.Add("r");
                    }

                });

                c1.SubscribeAsync("topic").Wait();

                await Task.Delay(500);

                c1.PublishAsync("topic").Wait();

                await Task.Delay(500);

                var flow = string.Join(string.Empty, events);
                Assert.AreEqual("cr", flow);

                try
                {
                    await testEnvironment.ConnectClientAsync(clientOptions);
                    Assert.Fail("same id connection is expected to fail");
                }
                catch
                {
                    //same id connection is expected to fail
                }

                await Task.Delay(500);

                flow = string.Join(string.Empty, events);
                Assert.AreEqual("cr", flow);

                c1.PublishAsync("topic").Wait();

                await Task.Delay(500);

                flow = string.Join(string.Empty, events);
                Assert.AreEqual("crr", flow);
            }
        }
```

3. See error:

* Assert.AreEqual("cr", flow); will fail with "crd" because the malicious connection will trigger a disconnect event in the server
* Assert.AreEqual("crr", flow); will fail with "crd" instead of "crr" because the authenticated client couldn't receive the message sent. If you do a c1.IsConnected call right there, it will return true and the disconnected handler for c1 won't be triggered. 

I created this test with a very simple example of authentication, where I only allow the first client with that Id to connect, but in my application I'm using username and password validation along with client id validation. If you use the same client_id, the disconnect event seems to be affecting the original connected user. 

### Expected behavior

The expectation is that if a malicious or rogue client tries to connect to the bus and it gets denied, the client that was already connected and authenticated should not be disturbed, which means, it should be allowed to send and receive messages. 

In my case, I expect exactly what is written on the test above. If the malicious client connect, the server won't trigger a disconnect event; the server won't clean up the connection for c1 and c1 will be able to send and receive messages after the malicious client tries to connect. 

I believe that formally disconnecting both is also a solution. So closing the connection for c1 is also possible outcome for me. 

### Screenshots
I don't have screenshoots

### Additional context / logging
If needed I can add more details

### Code example
Should be up there

**Comment by llrosa** [comment]
I have a fix proposal at https://github.com/chkr1011/MQTTnet/pull/786, if someone could review it and see if it makes sense. I don't believe that has any impact elsewhere, but it would be good if someone could take a closer look.

Thanks!

**Comment by SeppPenner** [comment]
I guess, this can be closed because #786 is already merged.
